src/types.ts

```diff
@@ -74,20 +75,19 @@ export type RawRequestOptions<V = Variables> = {
   signal?: Dom.RequestInit['signal']
 }
 
-export type RequestOptions<V = Variables> = {
-  document: RequestDocument
-  variables?: V
+export type RequestOptions<V = Variables, T = any> = {
+  document: RequestDocument | TypedDocumentNode<T, V>
   requestHeaders?: Dom.RequestInit['headers']
   signal?: Dom.RequestInit['signal']
-}
+} & (V extends never ? { variables?: never } : { variables: V })
 
 export type BatchRequestsOptions<V = Variables> = {
   documents: BatchRequestDocument<V>[]
   requestHeaders?: Dom.RequestInit['headers']
   signal?: Dom.RequestInit['signal']
 }
 
-export type RequestExtendedOptions<V = Variables> = { url: string } & RequestOptions<V>
+export type RequestExtendedOptions<V = Variables, T = any> = { url: string } & RequestOptions<V, T>
```

Comment:
Let's rename `T` to what it is.

Reply:
Hi @jasonkuhrt,

I'm working with @n1ru4l at The Guild, making this PR move forward.

Regarding your point, the `T` naming is used everywhere for the result type, do you want me to change it everywhere or just there?

Reply:
Hey, everywhere I guess.